ocpmodels/models/base.py

```diff
@@ -6,30 +6,219 @@
 """
 
 import logging
+from abc import abstractmethod
+from collections import defaultdict
 
 import torch
 import torch.nn as nn
+from e3nn import o3
 from torch_geometric.nn import radius_graph
 
 from ocpmodels.common.utils import (
+    cg_decomp_mat,
     compute_neighbors,
     conditional_grad,
     get_pbc_distances,
+    irreps_sum,
     radius_graph_pbc,
+    scatter_det,
 )
+from ocpmodels.models.gemnet_oc.layers.base_layers import Dense
 
 
 class BaseModel(nn.Module):
     def __init__(
-        self, num_atoms=None, bond_feat_dim=None, num_targets=None
+        self,
+        output_targets={},
+        node_embedding_dim=None,
+        edge_embedding_dim=None,
     ) -> None:
-        super(BaseModel, self).__init__()
-        self.num_atoms = num_atoms
-        self.bond_feat_dim = bond_feat_dim
-        self.num_targets = num_targets
+        super().__init__()
+
+        self.output_targets = output_targets
+        self.num_targets = len(output_targets)
+
+        self.module_dict = nn.ModuleDict({})
+        for target in output_targets:
+            if self.output_targets[target].get("custom_head", False):
+                if "irrep_dim" in self.output_targets[target]:
+                    if edge_embedding_dim is None:
+                        raise NotImplementedError(
+                            "Model does not support SO(3) equivariant prediction without edge embeddings."
+                        )
+                    embedding_dim = edge_embedding_dim
+                    output_shape = 1
+                else:
+                    embedding_dim = node_embedding_dim
+                    output_shape = self.output_targets[target].get("shape", 1)
+
+                layers = [
+                    Dense(
+                        embedding_dim,
+                        embedding_dim,
+                        activation="silu",
+                    )
+                ] * self.output_targets[target].get("num_layers", 2)
```

Comment:
Nit: Technically, this should be `num_layers - 1`, or the config name should change to `num_intermediate_layers`. Right now, setting num_layers=2 creates a 3 layer MLP.